### PR TITLE
[storage] Remove the now-useless sink handle

### PR DESCRIPTION
**Pull request by bkirwi**
We used to hold a handle in the sink to help "ride out" environmentd downtime. Critical handles now handle this case more definitively.

### Motivation

Previously-unspecified cleanup.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by bkirwi** [comment]
I kicked off a bunch of nightlies - at a glance all the failures are known issues or copy-to-s3 related, but I'll look into it properly before merging.

File: src/storage/src/storage_state.rs [hunk 1]
```diff
@@ -362,102 +355,6 @@ impl StorageInstanceContext {
     }
 }
 
-/// This maintains an additional read hold on the source data for a sink, alongside
-/// the controller's hold and the handle used to read the shard internally.
-/// This is useful because environmentd's hold might expire, and the handle we use
```
Comment (bkirwi) [inline src/storage/src/storage_state.rs:367]: FWIW, this is the bit that's no longer true... the critical handle will never expire, and it's responsible for taking the sink's reported frontier into account.

**Comment by bkirwi** [comment]
cc @aljoscha - this is the sketchy handle you mentioned in #26793!

**Review by aljoscha — APPROVED** [review]
Lhanks for taking the time! Looks good to me, if nightlies also agree

**Pull request merged**